dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp

```diff
@@ -34,7 +34,10 @@ PageDirectoryPtr PageDirectoryFactory::create(String storage_name, FileProviderP
     return createFromReader(storage_name, reader, std::move(wal));
 }
 
-PageDirectoryPtr PageDirectoryFactory::createFromReader(String storage_name, WALStoreReaderPtr reader, WALStorePtr wal)
+/// When create PageDirectory for dump snapshot, we should keep the last valid var_entry when it is deleted.
+/// Because there may be some upsert entry in later wal files, and we should keep the valid var_entry and the delete entry to delete the later upsert entry.
+/// And we don't restore the entries in blob store, because this PageDirectory is just read only for its entries.
```

Comment:
I think we should move these lines to `PageDirectory::tryDumpSnapshot`

Reply:
Fixed.